### PR TITLE
Fix #4620: Disable action button and add a spinner on top while send/swap unapproved tx is under construction. 

### LINKED ISSUE #4620: Swap transaction button sometimes slow to bring up the transaction confirmation screen leading to multiple taps and multiple transactions
### Description: 
Swap transaction button sometimes slow to bring up the transaction confirmation screen leading to multiple taps and multiple transactions

### Steps to Reproduce 
 
   1. Open Swap menu
   2. Enter token details
   3. Tap on swap, transaction confirmation doesn't show up which can lead to multiple taps in turn multiple confirm transaction popups

**Actual result:** 
Swap transaction button sometimes slow to bring up the transaction confirmation screen leading to multiple taps and multiple transactionsSwap transaction button sometimes slow to bring up the transaction confirmation screen leading to multiple taps and multiple transactions

**Expected result:**
Disable swap button when a tap is recorded so it avoids multiple taps and multiple confirmations

**Reproduces how often:** [Easily reproduced, Intermittent Issue]
-

**Brave Version:** 
1.33

**Device details:** 
All

**Website problems only:**
- did you check with Brave Shields down? NA
- did you check in Safari/Firefox (WkWebView-based browsers)?  NA

### Additional Information
cc: @kylehickinson @nuo-xu

### PR DESCRIPTION
**Pull request by nuo-xu**
## Summary of Changes
Created a new button that will have a spinner spinning and disabled when its loading. 

This pull request fixes #4620 

## Submitter Checklist:

- [x] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()`

## Reviewer Checklist:

- [ ] Issues include necessary QA labels:
  - `QA/(Yes|No)`
  - `bug` / `enhancement`
- [ ] Necessary [security reviews](https://github.com/brave/security/issues/new/choose) have taken place.
- [ ] Adequate unit test coverage exists to prevent regressions.
- [ ] Adequate test plan exists for QA to validate (if applicable).
- [ ] Issue and pull request is assigned to a milestone (should happen at merge time).

### CODE REVIEW

File: BraveWallet/WalletLoadingButton.swift [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    } label: {
+      ZStack {
+        title
+        Circle()
+          .trim(from: 0, to: 0.7)
+          .stroke(.white, lineWidth: 2)
+          .frame(width: length, height: length)
+          .rotationEffect(.degrees(isLoading ? 360 : 0))
+          .animation(.linear(duration: 0.5).repeatForever(autoreverses: false), value: isLoading)
+          .opacity(isLoading ? 1 : 0)
```
Comment (kylehickinson) [inline BraveWallet/WalletLoadingButton.swift]: Just use `ProgressView` instead?

File: BraveWallet/WalletLoadingButton.swift [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+import SwiftUI
+import BraveUI
+
+struct WalletLoadingButton<Title: View>: View {
+  @Binding var isLoading: Bool
+  var action: () -> Void
+  var title: Title
+  
+  @ScaledMetric private var length: CGFloat = 20.0
```
Comment (kylehickinson) [inline BraveWallet/WalletLoadingButton.swift]: Won't be needed with `ProgressView`

File: BraveWallet/Crypto/Stores/SwapTokenStore.swift [hunk 3]
```diff
@@ -208,12 +210,16 @@ public class SwapTokenStore: ObservableObject {
       let swapParams = swapParameters(for: .perSellAsset)
     else { return }
     
+    isMakingTx = true
     rpcController.network { [weak self] network in
       guard let self = self else {
         self?.state = .error(Strings.Wallet.unknownError)
         self?.clearAllAmount()
+        self?.isMakingTx = false
```
Comment (kylehickinson) [inline BraveWallet/Crypto/Stores/SwapTokenStore.swift]: These are not needed, if this guard fails then self is nil and won't execute any of this

File: BraveWallet/Crypto/Stores/SendTokenStore.swift [hunk 4]
```diff
@@ -150,8 +152,14 @@ public class SendTokenStore: ObservableObject {
     let weiFormatter = WeiFormatter(decimalFormatStyle: .decimals(precision: 18))
     guard let token = selectedSendToken, let weiHexString = weiFormatter.weiString(from: amount, radix: .hex, decimals: 18) else { return }
     
+    isMakingTx = true
     rpcController.network { [weak self] network in
-      guard let self = self else { return }
+      guard let self = self else {
+        self?.isMakingTx = false
+        return
+      }
```
Comment (kylehickinson) [inline BraveWallet/Crypto/Stores/SendTokenStore.swift]: These are not needed, if this guard fails then self is nil and won't execute any of this

File: BraveWallet/WalletLoadingButton.swift [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+  
+  init(
+    isLoading: Binding<Bool>,
```
Comment (kylehickinson) [inline BraveWallet/WalletLoadingButton.swift]: This does not need to be a `Binding` as you are not setting `isLoading` from within the view

File: BraveWallet/Crypto/BuySendSwap/SwapCryptoView.swift [hunk 6]
```diff
@@ -376,13 +376,17 @@ struct SwapCryptoView: View {
...
+          WalletLoadingButton(
+            isLoading: $swapTokensStore.isMakingTx,
+            action: {
+              swapTokensStore.prepareSwap()
+            },
+            title: {
+              Text(swapButtonTitle)
+            }
+          )
+            .disabled(isSwapButtonDisabled || swapTokensStore.isMakingTx)
```
Comment (kylehickinson) [inline BraveWallet/Crypto/BuySendSwap/SwapCryptoView.swift]: Add `swapTokensStore.isMakingTx` check to `isSwapButtonDisabled` logic

File: BraveWallet/Crypto/BuySendSwap/SendTokenView.swift [hunk 7]
```diff
@@ -121,19 +121,23 @@ struct SendTokenView: View {
...
+              title: {
+                Text(Strings.Wallet.sendCryptoSendButtonTitle)
               }
-            }) {
-              Text(Strings.Wallet.sendCryptoSendButtonTitle)
-            }
+            )
             .buttonStyle(BraveFilledButtonStyle(size: .normal))
-            .disabled(isSendDisabled)
+            .disabled(isSendDisabled || sendTokenStore.isMakingTx)
```
Comment (kylehickinson) [inline BraveWallet/Crypto/BuySendSwap/SendTokenView.swift]: Add `sendTokenStore.isMakingTx` check to `isSendDisabled ` logic

File: BraveWallet/WalletLoadingButton.swift [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+  }
+  
+  var body: some View {
+    Button {
+      if !isLoading {
+        action()
+      }
+    } label: {
+      ZStack {
+        title
```
Comment (kylehickinson) [inline BraveWallet/WalletLoadingButton.swift]: So wait is `title` staying visible when its loading? Shouldn't it hide?

File: BraveWallet/WalletLoadingButton.swift [hunk 9]
```diff
@@ -0,0 +1,63 @@
+// Copyright 2021 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import SwiftUI
+import BraveUI
+
+struct WalletLoadingButton<Title: View>: View {
```
Comment (kylehickinson) [inline BraveWallet/WalletLoadingButton.swift]: `Title` should be `Label` to match SwiftUI `Button`

**Review by kylehickinson — CHANGES_REQUESTED** [review]

File: BraveWallet/WalletLoadingButton.swift (hunk 1 above)
Comment (nuo-xu) [reply BraveWallet/WalletLoadingButton.swift]: I was trying to make it look the same as Desktop.

Comment (kylehickinson) [reply BraveWallet/WalletLoadingButton.swift]: We'd probably want to use a custom `ProgressViewStyle` in that case but I'd say for now just replace with `ProgressView` and we can revisit later

Comment (nuo-xu) [reply BraveWallet/WalletLoadingButton.swift]: ok. will update.

**Review by kylehickinson — APPROVED** [review]

**Pull request merged**